src/engine/strat_engine/thinpool.rs

```diff
@@ -197,34 +330,40 @@ impl ThinPool {
         Ok(())
     }
 
-    /// Get an immutable reference to the sparse segments of the ThinPool.
-    pub fn spare_segments(&self) -> &[Segment] {
-        &self.meta_spare
-    }
-
-    /// The segments belonging to the thin pool meta device.
-    pub fn thin_pool_meta_segments(&self) -> &[Segment] {
-        self.thin_pool.meta_dev().segments()
-    }
-
-    /// The segments belonging to the thin pool data device.
-    pub fn thin_pool_data_segments(&self) -> &[Segment] {
-        self.thin_pool.data_dev().segments()
-    }
-
-    /// The segments belonging to the MDV.
-    pub fn thin_pool_mdv_segments(&self) -> &[Segment] {
-        self.mdv.segments()
-    }
-
     /// Get the devicemapper::ThinPoolDev for this pool. Used for testing.
     pub fn thinpooldev(&self) -> &ThinPoolDev {
         &self.thin_pool
     }
 
     /// Extend the thinpool with new data regions.
-    pub fn extend_data(&mut self, dm: &DM, segs: Vec<Segment>) -> EngineResult<()> {
-        Ok(self.thin_pool.extend_data(dm, segs)?)
+    pub fn extend_data(&mut self, dm: &DM, new_segs: Vec<BlkDevSegment>) -> EngineResult<()> {
+        self.thin_pool.extend_data(dm, map_to_dm(&new_segs))?;
```

Comment:
Move this call to end of method? It would need to be to a method other than ```extend_data```, one that reset the data. That is kind of scary, obviously, but we should be prepared to be able to stand by the correctness of this code. I think such an approach would be more robust to maintain.

The drawback is that you could coalesce the segments in this method, and then fail, and then, ideally, do the operation of uncoalescing them, which would be a pain. But that can be solved by making a local copy of the coalesced segments, doing the coalescing, making the call with the copy, and then setting the field.

Reply:
I think we want this at the top, so we don't update our state until we know the lower-level operation has succeeded.

Reply:
My point is that actually changing the state should not be necessary, especially now that segments can be cloned.

Reply:
I'm still not comprehending an alternative implementation that is better. Maybe it'd be faster if you can add a commit to this PR that demonstrates?

Reply:
I will give it a try.